client/app/scripts/components/zoomable-canvas.js

```diff
@@ -144,64 +133,122 @@ class ZoomableCanvas extends React.Component {
   }
 
   updateZoomLimits(props) {
-    const zoomLimits = props.layoutZoomLimits.toJS();
-
-    this.zoom = this.zoom.scaleExtent([zoomLimits.minScale, zoomLimits.maxScale]);
-
-    if (props.bounded) {
-      this.zoom = this.zoom
-        // Translation limits are only set if explicitly demanded (currently we are using them
-        // in the resource view, but not in the graph view, although I think the idea would be
-        // to use them everywhere).
-        .translateExtent([
-          [zoomLimits.minTranslateX, zoomLimits.minTranslateY],
-          [zoomLimits.maxTranslateX, zoomLimits.maxTranslateY],
-        ])
-        // This is to ensure that the translation limits are properly
-        // centered, so that the canvas margins are respected.
-        .extent([
-          [props.canvasMargins.left, props.canvasMargins.top],
-          [props.canvasMargins.left + props.width, props.canvasMargins.top + props.height]
-        ]);
-    }
-
-    this.setState(zoomLimits);
+    this.setState(props.layoutLimits.toJS());
   }
 
   // Restore the zooming settings
   restoreZoomState(props) {
     if (!props.layoutZoomState.isEmpty()) {
       const zoomState = props.layoutZoomState.toJS();
 
-      // After the limits have been set, update the zoom.
-      this.svg.call(this.zoom.transform, zoomIdentity
-        .translate(zoomState.translateX, zoomState.translateY)
-        .scale(zoomState.scaleX, zoomState.scaleY));
-
       // Update the state variables.
       this.setState(zoomState);
       this.zoomRestored = true;
     }
   }
 
   canChangeZoom() {
-    const { disabled, layoutZoomLimits } = this.props;
-    const canvasHasContent = !layoutZoomLimits.isEmpty();
+    const { disabled, layoutLimits } = this.props;
+    const canvasHasContent = !layoutLimits.isEmpty();
     return !disabled && canvasHasContent;
   }
 
-  zoomed() {
+  handlePanStart() {
+    this.setState({ isPanning: true });
+  }
+
+  handlePanEnd() {
+    this.setState({ isPanning: false });
+  }
+
+  handlePan() {
+    let state = { ...this.state };
```

Comment:
Good catch! Although note that this has trivial complexity compared to all the rerendering stuff that follows each panning event :)